Support non ascii characters in the source code

yukoba opened the pull request:
Ray throws an exception when non ascii characters are used in the source code.

Sample code. I used Japanese in the comment.

```python
import time
import ray

@ray.remote
def f():
    # 日本語
    time.sleep(1)
    return 1

ray.init()
results = ray.get([f.remote() for i in range(4)])
print(results)
```

Raised exception.

```
Traceback (most recent call last):
  File "raytest.py", line 4, in <module>
    @ray.remote
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 2382, in remote
    return make_decorator(worker=worker)(args[0])
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 2291, in decorator
    num_return_vals, max_calls)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/remote_function.py", line 44, in __init__
    self._function_descriptor = FunctionDescriptor.from_function(function)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/ray/function_manager.py", line 126, in from_function
    source = inspect.getsource(function).encode("ascii")
UnicodeEncodeError: 'ascii' codec can't encode characters in position 27-29: ordinal not in range(128)
```

pcmoritz commented:
Good catch! Can you also add a small test for it?

yukoba commented:
Where should I add test? Should I add here?
https://github.com/ray-project/ray/blob/master/test/mini_test.py#L29

By the way, this project uses ascii many places, but I think it should all be replaced to utf-8.

pcmoritz commented:
The test should be added here: https://github.com/ray-project/ray/blob/master/test/runtest.py

yukoba commented:
I added a test.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11929/
Test PASSed.

raulchen commented on test/runtest.py:
```diff
@@ -2767,3 +2767,14 @@ def f():
+
+
+def test_non_ascii_comment(shutdown_only):
+    ray.init(num_cpus=1)
```
Nit, you can change `shutdown_only` to `ray_start`, then `ray.init(num_cpus=1)` can be removed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11932/
Test FAILed.

yukoba commented:
I changed to ```ray_start```.
Is Jenkins failed?

yukoba commented:
I also add ```# coding: UTF-8``` to runtest.py for Python 2.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11942/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11943/
Test FAILed.

yukoba commented:
Travis failed at Python 2.7.

```
test/runtest.py::test_non_ascii_comment FAILED                           [100%]
=================================== FAILURES ===================================
____________________________ test_non_ascii_comment ____________________________
ray_start = None
    def test_non_ascii_comment(ray_start):
>       @ray.remote
        def f():
test/runtest.py:2774: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../.local/lib/python2.7/site-packages/ray-0.7.0.dev0-py2.7-linux-x86_64.egg/ray/worker.py:2515: in remote
    return make_decorator(worker=worker)(args[0])
../../../.local/lib/python2.7/site-packages/ray-0.7.0.dev0-py2.7-linux-x86_64.egg/ray/worker.py:2424: in decorator
    num_return_vals, max_calls)
../../../.local/lib/python2.7/site-packages/ray-0.7.0.dev0-py2.7-linux-x86_64.egg/ray/remote_function.py:44: in __init__
    self._function_descriptor = FunctionDescriptor.from_function(function)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cls = <class 'ray.function_manager.FunctionDescriptor'>
function = <function f at 0x7fefd84e5848>
    @classmethod
    def from_function(cls, function):
        """Create a FunctionDescriptor from a function instance.
    
        This function is used to create the function descriptor from
        a python function. If a function is a class function, it should
        not be used by this function.
    
        Args:
            cls: Current class which is required argument for classmethod.
            function: the python function used to create the function
                descriptor.
    
        Returns:
            The FunctionDescriptor instance created according to the function.
        """
        module_name = function.__module__
        function_name = function.__name__
        class_name = ""
    
        function_source_hasher = hashlib.sha1()
        try:
            # If we are running a script or are in IPython, include the source
            # code in the hash.
>           source = inspect.getsource(function).encode("utf-8")
E           UnicodeDecodeError: 'ascii' codec can't decode byte 0xe6 in position 39: ordinal not in range(128)
../../../.local/lib/python2.7/site-packages/ray-0.7.0.dev0-py2.7-linux-x86_64.egg/ray/function_manager.py:126: UnicodeDecodeError
```

yukoba commented:
I think I fixed for Python 2.

hashlib update() accepts following arguments.
- Python 2: str
- Python 3: bytes

Originally, ```inspect.getsource(function).encode("ascii")``` was used to convert this.
- Python 2: str -> str
- Python 3: str -> bytes

It was doing nothing in Python 2.
However, in Python 2, ```str.encode("utf-8")``` raises UnicodeDecodeError for a utf-8 str.

So I changed to the following code to support both Python 2 and 3.
```python
            source = inspect.getsource(function)
            if sys.version_info[0] >= 3:
                source = source.encode()
```

```str.encode()``` is not necessary for Python 2.
And the argument of encode() is not necessary for Python 3 because the default is utf-8.

raulchen commented:
Cool. BTW, you can change it to `source = six.ensure_binary(inspect.getsource(function))`, which is a little bit nicer.

yukoba commented:
```six.ensure_binary()``` is a new six 1.12 API.
It is not in the old six 1.10 document. https://pythonhosted.org/six/ 

Ray depends on six 1.0 and I think I have to modify this too.
https://github.com/ray-project/ray/blob/master/python/setup.py#L146

Can I keep my code? I already uninstalled Python 2 and I feel there is no big difference.

raulchen commented:
ok, that's fine

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11982/
Test FAILed.

yukoba commented:
Travis failed but I feel it is not related to my pull request.

robertnishihara commented:
Also related to https://github.com/ray-project/ray/issues/2380, though https://github.com/ray-project/ray/issues/2380 probably still doesn't work after this PR.

yukoba commented:
I think they come from here.
https://github.com/ray-project/ray/blob/master/python/ray/function_manager.py#L229
https://github.com/ray-project/ray/blob/master/python/ray/function_manager.py#L250
https://github.com/ray-project/ray/blob/master/python/ray/function_manager.py#L593

If you don't care about Python 2, 
all the ```str.encode("ascii")``` should be replaced to ```str.encode()```.
Or ```six.ensure_binary(str)``` if you can use latest six 1.12.

yukoba commented:
I think there are 3 options to avoid ```str.encode("ascii")```.
1. Ignore Python 2 and replace it to ```str.encode()```
2. Use latest six 1.12 ```ensure_binary(str)```.
3. Because six 1.12 is too new, copy and paste ```ensure_binary(str)``` and use it. https://github.com/benjaminp/six/blob/d927b9e27617abca8dbf4d66cc9265ebbde261d6/six.py#L853

raulchen commented:
Travis failure isn't related.  #2380 should still fail because this PR only fixes non-ascii comments, not non-ascii function names. But I think it's okay to merge this first. In a subsequent RP, I suggest updating six and fix everything with `six.ensure_binary/ensure_str`

yukoba commented:
Thank you for merging.

kyang922 commented:
this issue is not limited in non-ascii comments or function names, any non-ascii characters in the function source will cause this error.

```
import ray

@ray.remote
class A:
    def a(self):
        print("あ")
        
ray.init()
A.remote()
```
will also give
`
UnicodeEncodeError: 'ascii' codec can't encode character '\u3042' in position 257: ordinal not in range(128)
`

it seems that your code had been updated to use dis.dis instead of inspect.getsource, but this problem still remain not fixed since you still using string_file.getvalue().encode("ascii") in [function_manager.py](https://github.com/ray-project/ray/blob/ca6eabc9cb728a829d5305a4dd19216ed925f2e4/python/ray/function_manager.py#L123)

your code is doing like this
```
from io import StringIO
import dis

string_file = StringIO()

def a():
    print("あ")
    
dis.dis(a, file=string_file)
string_file.getvalue().encode("ascii")
```
and it will give the error above.  I don't know why you stick with using 'ascii' here, but it will block out all users who not only use 'ascii' chars.

and about the non-ascii function names, it's the problem of python 2, since it not support that. so there is no need to get it fixed.

but I really need the non-ascii support in the code get supported since I am developing a application on the ray. or can you tell me which version shall me downgrade to that can support it?
